### ISSUE TITLE
lightgamma, texgamma bug

### ISSUE THREAD

**Issue by hotactions** [ValveSoftware/halflife#232]
When you change lightgamma cvar you always get killed or write more than 1 time you get client crashed this should be fixed.

also texgamma sometimes crashing client

**Comment by mikela-valve** [comment]
Hi all, I've made a change in the latest CS beta that disables non-default lightgamma and texgamma altogether seeing as how even with sv_cheats set you are still killed after adjusting the cvars, so I think the intent was for them to not be changeable for CS. If this is a problem I can look into some way for the server to make the decision as to whether lightgamma is settable or not.

**Comment by AnAkkk** [comment]
Sounds good!

**Comment by drogga** [comment]
But now you can't adjust texgamma at all! Allow a range of 2 - 3 for texgamma. On some monitors, the standard value looks just awful! Fixed by setting texgamma 2.5 value

**Comment by mikela-valve** [comment]
@drogga Unfortunately I don't think I can unlock the value of texgamma, though I'd be up to setting a different default value if there is some consensus of a better one.

I did some testing around it with different values and it appears that only some of the lighting updates dynamically when changing texgamma, it mostly affects precomputed texture gamma and so needs to be set before textures are loaded. See these images:

texgamma = 1.8 (the minimum)
![DFE29F33-B74A-41D4-92AA-1C6B0D71BF31](https://user-images.githubusercontent.com/49570952/57719347-67997d00-7634-11e9-996b-6805ef5f5133.jpg)

texgamma = 3.0 (the maximum)
![BFD64AA0-A7ED-414B-B411-7BC9F223BABE](https://user-images.githubusercontent.com/49570952/57719354-6c5e3100-7634-11e9-83e1-dc79fac91a79.jpg)

**Comment by drogga** [comment]
Maybe I found a compromise solution for "texgamma". Suffice it to disable change this option in game, leaving the possibility to set the value only once, through cfg. This will solve all the problems. Another similar way - remove option "texgamma" from the console, but instead add a startup parameter "-texgamma" or "-tex" (eg "-tex 2.5") - it's just solve the problem by making it impossible to change a texgamma already in the game (when changing the card).

texgamma allows you to choose the right value for graphics settings. It is also useful when creating a gaming video. Completely disable it is not desirable.

Checkin this:

1) gamma 2.5; brightness 2; texgamma 2
![tex1](https://user-images.githubusercontent.com/43634578/57731463-89542d80-764f-11e9-9e8d-cdde1a12333e.PNG)
Normal balanced display (default)

2) gamma 3.0; brightness 2; texgamma 2
![tex2](https://user-images.githubusercontent.com/43634578/57731479-94a75900-764f-11e9-8420-c5bb685967a8.PNG)
Whitish textures, completely wrong

3) gamma 3.0; brightness 2; texgamma 2.5
![tex3](https://user-images.githubusercontent.com/43634578/57731496-9ec95780-764f-11e9-86eb-1e2b02e37e61.PNG)
Already acceptable!

I think that with an increase in the value of GAMMA, TEXGAMMA must also be raised to maintain balance.
With gamma 2.5 - texgamma 2.0; with gamma 3.0 - texgamma 2.5

The only complaint is the possibility of abuse "texgamma" - is achieved by changing the value in an already running game (if originally stood for "0", then we went to the server and changed to "3"), then when changing maps, models are a bit lighter:

![1](https://user-images.githubusercontent.com/43634578/57731531-b7d20880-764f-11e9-9bda-bcff3aa4edcc.PNG)

![2](https://user-images.githubusercontent.com/43634578/57731549-c15b7080-764f-11e9-9cb1-f1757419035e.PNG)

This is a bug that, in a good way, should just be fixed. But if it is very difficult to fix, it is enough to block the change of texgamma in the game and disable values ​​less than 2.0 and more than 3.0. As well as that you could ask texgamma only once before starting the game (start parameters or through the CFG, as I wrote above). This will solve all the problems.

CS 1.6 not rich in image settings, I think shouldn’t take away the latest customization options from players and moviemakers.

**Comment by AnAkkk** [comment]
Except that texgamma is not supposed to be changed, before or during a game - a lot of people would see it as an unfair advantage, and I am pretty sure that using non default values of this cvar was actually banned from all competitions. e.g. you can find it in IEM rulebook.

**Comment by mikela-valve** [comment]
@AnAkkk I actually did recently get to adding synchronization of `sv_cheats` between the client and server so I'll be returning this back to its original logic of allowing `texgamma` and `lightgamma` to be changed if the server is running with `sv_cheats` enabled. This should be the be the best compromise to still allow them to be usable for non-competitive play at least.

**Comment by AnAkkk** [comment]
@mikela-valve: would that still not cause issues because when you change texgamma and lightgamma it does not reload the textures, so even if they get forced to different values due to sv_cheats, it wouldn't undo modifications they did to the game?

**Comment by mikela-valve** [comment]
@AnAkkk You're right, I was thinking that enough of the lighting was calculated during a map load (so it would be enough for the server to change the value on the client during initial connection) but it appears that a full client reload is currently necessary to recompute that. 

There are some workarounds I could implement as a roundabout way of enforcing this like preventing a client that has changed its lightgamma/texgamma values since startup from connecting to a server that doesn't have `sv_cheats` set but for now it's probably best to leave these values fixed and see how that goes for the community.

Thanks!

**Comment by mikela-valve** [comment]
Closing this as fixed, we can open a new issue in the future if necessary for determining how to re-enable some things if possible.
